polkadot/node/network/statement-distribution/src/v2/mod.rs

```diff
@@ -2126,17 +2135,54 @@ async fn provide_candidate_to_grid<Context>(
 	}
 }
 
-fn group_for_para(
+// Utility function to populate per relay parent `ParaId` to `GroupIndex` mappings.
+fn determine_groups_per_para(
 	availability_cores: &[CoreState],
 	group_rotation_info: &GroupRotationInfo,
-	para_id: ParaId,
-) -> Option<GroupIndex> {
-	// Note: this won't work well for on-demand parachains as it assumes that core assignments are
-	// fixed across blocks.
-	let core_index = availability_cores.iter().position(|c| c.para_id() == Some(para_id));
+	maybe_claim_queue: &Option<ClaimQueueSnapshot>,
+	max_candidate_depth: usize,
+) -> HashMap<ParaId, Vec<GroupIndex>> {
+	// Determine the core indices occupied by each para at the current relay parent. To support
+	// on-demand parachains we also consider the core indices at next block if core has a candidate
+	// pending availability.
+	let para_core_indices = availability_cores.iter().enumerate().filter_map(|(index, core)| {
+		match core {
+			CoreState::Scheduled(scheduled_core) =>
+				Some((scheduled_core.para_id, CoreIndex(index as u32))),
+			CoreState::Occupied(occupied_core) => {
+				if max_candidate_depth >= 1 {
+					// Use claim queue if available, or fallback to `next_up_on_available`
+					let maybe_scheduled_core = match maybe_claim_queue {
+						Some(claim_queue) => {
+							// What's up next on this core ?
+							fetch_next_scheduled_on_core(claim_queue, CoreIndex(index as u32))
+						},
+						None => {
+							// Runtime doesn't support claim queue runtime api. Fallback to
+							// `next_up_on_available`
+							occupied_core.next_up_on_available.clone()
+						},
+					};
+
+					maybe_scheduled_core
+						.filter(|scheduled_core| scheduled_core.para_id == occupied_core.para_id())
+						.map(|scheduled_core| (scheduled_core.para_id, CoreIndex(index as u32)))
+				} else {
+					None
+				}
+			},
+			CoreState::Free => None,
+		}
+	});
 
-	core_index
-		.map(|c| group_rotation_info.group_for_core(CoreIndex(c as _), availability_cores.len()))
+	let mut groups_per_para = HashMap::new();
+	// Map from `CoreIndex` to `GroupIndex` and collect as `HashMap`.
+	for (para, core_index) in para_core_indices {
+		let group_index = group_rotation_info.group_for_core(core_index, availability_cores.len());
```

Comment:
Group rotation info needs the block to be able to compute the index with the usage of ClaimQueue which I guess is from the future, does this gives us the correct result at rotation boundaries ?

Reply:
Here we are interested in the group mapping for a core at the relay parent which is supposed to back candidates built on this relay parent. The claim queue surely refers to something in the future, but that backing group is always determined by the relay parent used to build the candidate.

Reply:
So yeah, it should give the correct result when we pass the rotation boundary, but that would be at the next relay parent.